move has_rotten_away from map to item

KorGgenT opened the pull request:
<!--
### How to use
Leave the headings unless they don't apply to your PR, replace commented out text (surrounded with <!–– and ––>) with text describing your PR.
NOTE: Please grant permission for repository maintainers to edit your PR.
It is EXTREMELY common for PRs to be held up due to trivial changes being requested and the author being unavailable to make them.
-->

#### Summary
<!--
A one-line description of your change that will be extracted and added to the [project changelog](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/changelog.txt).

The format is (ignore the square brackets): ```SUMMARY: [Category] "[description]"```

The categories to choose from are:

* Features
* Content
* Interface
* Mods
* Balance
* Bugfixes
* Performance
* Infrastructure
* Build
* I18N

Example: ```SUMMARY: Content "Adds new mutation category 'Mouse'"```

See the [Changelog Guidelines](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/CHANGELOG_GUIDELINES.md) for explanations of the categories.
-->
```SUMMARY: Infrastructure "move has_rotten_away() from map to item"```
#### Purpose of change
<!--
If there's an existing issue describing the problem this PR addresses or the feature it adds, please link it like: ```#1234```
If it *fully* resolves an issue, link it like: Fixes #1234
Even if the issue describes the problem, please provide a few-sentence summary here.
Example: Fixes #1234 - XL mutants cannot wear arm/leg splints due to missing OVERSIZE flag.
If there is no related issue, please describe the issue you are addressing, including how to trigger a bug if this is a bugfix.
Don't put the backticks around the `#` and issue or pull request number to allow the GitHub automatically reference to it.
-->
This function really shouldn't have been in map. there was literally no reason for it. when i moved it to item, it didn't use any functions or variables from map at all! The main reason i'm moving it is because of the item contents refactor i'm working on. map really shouldn't be accessing item contents directly.
#### Describe the solution
<!--
How does the feature work, or how does this fix a bug?
The easier you make your solution to understand, the faster it can get merged.
-->
move the function to the item scope. this changes the parameter to just `const tripoint &pos` moved to item.cpp. it was only used in one place so there wasn't any extra cleanup.
#### Testing
<!--
Describe what steps you took to test that this PR resolved the bug or added the feature, and what tests you performed to make sure it didn't cause any regressions.
Also include testing suggestions for reviewers and maintainers.
-->
compiled.

BevapDin commented on src/item.h:
```diff
@@ -725,6 +725,14 @@ class item : public visitable<item>
         void mod_charges( int mod );
+        /**
+         * Whether the item has to be removed as it has rotten away completely.
+         * @param itm Item to check for rotting
```
```suggestion
```

BevapDin commented on src/item.h:
```diff
@@ -725,6 +725,14 @@ class item : public visitable<item>
+        /**
+         * Whether the item has to be removed as it has rotten away completely.
+         * @param itm Item to check for rotting
+         * @param pnt The *absolute* position of the item in the world (not just on this map!),
```
```suggestion
         * @param pnt The *absolute* position of the item in the world (see @ref map::getabs),
```

BevapDin commented on src/item.h:
```diff
@@ -725,6 +725,14 @@ class item : public visitable<item>
          */
         void mod_charges( int mod );
+        /**
+         * Whether the item has to be removed as it has rotten away completely.
```
You may want to add a note that this calls `item::process_temperature_rot` and thereby modifies the item. Currently it sounds as if this only query some condition.

Hirmuolio commented:
`has_rotten_away()` and `has_rotten_away()` being separate things that work together to kind of do the same thing is pretty confusing too.

Also now that I look at this the reason why items in freezer sometimes rot away may be due to the `has_rotten_away()`. This thing only uses item and item location and does not include the temperature_flag which marks the container temperature. So if `has_rotten_away()` is called on an item that is in freezer/fridge/etc it will process the item as if it was on ground.

Hirmuolio commented:
Acutally I now think that `map::has_rotten_away( item &itm, const tripoint &pnt )` doesn't need to exist at all.

Just check `has_rotten_away()` at the end of `item::process_temperature_rot`. Remove items that have rotten away and spawn carrion spawns as needed.

The original idea was probably to process food items once when they are loaded and after that limit how often they are processed with `processing_speed`.  
But that limit doesn't seem to work and food items are processed on every turn. This doesn't cause problems because there is another time gate in process_temperature_rot that bails out early if the item doesn't need processing yet.

KorGgenT commented:
> Acutally I now think that map::has_rotten_away( item &itm, const tripoint &pnt ) doesn't need to exist at all.

right, that's why i removed it from map and moved it to item... i'm having a hard time following what you're saying though. maybe you could follow up this pr to do what you said?

Hirmuolio commented:
Currently this seems to happen:
* Item is loaded to reality bubble.
* `map::remove_rotten_items` is called.
* It calls `map::has_rotten_away`on all items on ground (so this can't be the reason for freezer problem).
* it calls `item::process_temperature_rot`.
* Return back to `map::remove_rotten_items`  and and call `map::rotten_item_spawn` if item rotted away.

Then on the next turn the item is procesed normally and `item::process_temperature_rot` is done normally making the whole ordeal kind of pointless.

So we could just remove that part and do:

* Item is loaded to reality bubble.
* Do no special temperature/rot processing on it.
* On next turn it is processed.
* At the end of process_temperature_rot check wether it has rotten away and remove the item if it has rotten away and call `map::rotten_item_spawn`.
